Dynamic Inventory Management System

Paras-29 opened the pull request:
Successfully Complete the code for Dynamic Inventory Management System.

I were tasked with building a Dynamic Inventory Management System for a shop. 
The shop sells multiple products, but the number of products is not known in advance. 
Therefore, you must use dynamic memory allocation to manage product details 
efficiently. 
Each product has the following attributes: 
• Product ID (integer) 
• Product Name (string) 
• Price (float) 
• Quantity (integer)

Paras-29 committed:
Successfully write the code for Inventory System

Paras-29 committed:
Remove sonarImages.c

badal-soni commented on sonarImages.c:
```diff
@@ -1,169 +0,0 @@
-#include <stdio.h>
```
why you've deleted this file ?

Paras-29 replied:
By Mistake this file also push in this branch that's why I deleted this.

badal-soni replied:
Then after merging this file will also get deleted. Undo this change. THIS IS NOT EXPECTED IN WEEK-3.

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+bool isValidRangeInt(int val, int min, int max);
+bool isValidRangeFloat(float val, float min, float max);
+bool isDuplicateId(Product *products, int numOfProduct, int id);
+
```
remove extra one line space

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+
+
+int main() {
+
```
remove one line extra space. Give one line space between methods and no space at start and end in each method.

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+    int numOfProducts;
+    Product *products = NULL;
+
+    do{
```
```suggestion
    do {
```

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+
+        if(!isValidRangeInt(numOfProducts, MIN_PRODUCTS, MAX_PRODUCTS)) {
+            printf("Invalid number of products. Must be between %d and %d.\n", MIN_PRODUCTS, MAX_PRODUCTS);
+        }
```
don't you think we should break in the else case ?

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+        printf("Enter initial number of products: ");
+        scanf("%d", &numOfProducts);
+
+        if(!isValidRangeInt(numOfProducts, MIN_PRODUCTS, MAX_PRODUCTS)) {
```
```suggestion
        if (!isValidRangeInt(numOfProducts, MIN_PRODUCTS, MAX_PRODUCTS)) {
```

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+            printf("Invalid number of products. Must be between %d and %d.\n", MIN_PRODUCTS, MAX_PRODUCTS);
+        }
+
+    }while((!isValidRangeInt(numOfProducts, MIN_PRODUCTS, MAX_PRODUCTS)));
```
```suggestion
    } while ((!isValidRangeInt(numOfProducts, MIN_PRODUCTS, MAX_PRODUCTS)));
```

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+       for (int i = 0; i < numOfProduct; i++) {
+            printf("\nEnter details for product %d:\n", i + 1);
+
+        do {
```
fix the indentation block and move the logic of this for loop into a separate function.

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+                printf("Invalid product id. Must be between %d and %d.\n", MIN_ID, MAX_ID);
+            }
+            else if (isDuplicateId(products, i, products[i].id)) {
+            printf("Duplicate ID! Please enter a unique product ID.\n");
```
fix indentation

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+        while (getchar() != '\n');
+        printf("Product Name: ");
+        fgets(products[i].name, NAME_LEN, stdin);
+        products[i].name[strcspn(products[i].name, "\n")] = '\0';
```
create your own function instead of using the build in one. Do this for all the places where you've used standard functions and can be developed easily.

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+
+void addProduct(Product **products, int *numOfProduct) {
+
+    if (*numOfProduct >= MAX_PRODUCTS) {
```
Shouldn't we check the available space in the array ? Your current logic will allow the user to add products even larger than the defined limit.

Paras-29 replied:
I don't understand your comment correctly because in the add product function we are reallocating the space for new product in the constraints there no limit to add the product. Limit is only for the initial product that is 100.

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+        printf("Product ID: ");
+        scanf("%d", &newProduct->id);
+
+        if((!isValidRangeInt(newProduct->id, MIN_ID, MAX_ID)) ){
```
I am not able to understand why are you adding extra brackets everywhere ? Are you reviewing your code by yourself before sending it for review ?

Paras-29 replied:
its mistake

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+                printf("Invalid product id. Must be between %d and %d.\n", MIN_ID, MAX_ID);
+        }
+        else if (isDuplicateId(*products, *numOfProduct, newProduct->id)) {
+        printf("Duplicate ID! Please enter a unique product ID.\n");
```
fix indentation

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+                printf("Enter new Quantity: ");
+                scanf("%d", &newQty);
+
+                if((!isValidRangeInt(newQty, MIN_QUANTITY, MAX_QUANTITY))){
```
fix indentation issue everywhere.

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+
+    printf("Products Found:\n");
+    for (int i = 0; i < numOfProduct; i++) {
+        if (strstr(products[i].name, keyword) != NULL) {
```
implement your own logic to check if the input is a substring or not

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+    if (!isFound){
+        printf("No matching products found.\n");
+    }    
+
```
remove 2 line spaces

Paras-29 replied:
okk

badal-soni commented on inventory.c:
```diff
@@ -0,0 +1,389 @@
+
+        while (getchar() != '\n');
+        printf("Product Name: ");
+        fgets(products[i].name, NAME_LEN, stdin);
```
should also check if the product with same name (ignoring case) exists in the array.

Paras-29 replied:
okk

Paras-29 committed:
Resolve all the comments

badal-soni reviewed (changes requested):
still you're deleting the sonarImages.c file ? May I know why ?

Paras-29 commented:
I deleted the sonarImage.c file from this branch because it is not for this week assignment
So they is no need of this file and this file is already in the main branch, I deleted this file from this branch only , they is no need of this file here and by mistake I push this file on this branch also so I deleted this file.

badal-soni commented:
> I deleted the sonarImage.c file from this branch because it is not for this week assignment So they is no need of this file and this file is already in the main branch, I deleted this file from this branch only , they is no need of this file here and by mistake I push this file on this branch also so I deleted this file.

If you would know how git works then you would not have added this comment.
